<a href="https://colab.research.google.com/github/CompPsychology/psych290_colab_public/blob/main/notebooks/week-03/W3_Tutorial_05_DLATK_lexiconExtraction_mostFreqWords_(dla_tutorial)_withSolutions.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# W3 Tutorial 5 -- Lexicon tables & feature extraction (DB: dla_tutorial) (2025-03)

(c) Johannes Eichstaedt & the World Well-Being Project, 2023.

✋🏻✋🏻 NOTE - You need to create a copy of this notebook before you work through it. Click on "Save a copy in Drive" option in the File menu, and safe it to your Google Drive.

✉️🐞 If you find a bug/something doesn't work, please slack us a screenshot, or email johannes.courses@gmail.com.

## 1) Setting up Colab with DLATK and SQLite

If colab asks you about this not being authored by Google, say "Run anyway."

### 1a to 1c) Streamlined: Setting up Colab with DLATK and your data

In [ ]:
# assigning the database name
database = "dla_tutorial"

In [ ]:
########### 1a) Install

# installing DLATK and necessary packages
!git clone -b psych290 https://github.com/dlatk/dlatk.git
!pip install -r dlatk/install/requirements.txt
!pip install dlatk/
!pip install wordcloud langid jupysql

########### 1b) Download data and insert into SQLite database

# this download the csvs we need for this tutorial
!git clone https://github.com/CompPsychology/psych290_data.git

# load the required package -- similar to library() function in R
import os
from dlatk.tools.importmethods import csvToSQLite

# store the complete path to the database -- sqlite_data/[database_name].db
database_path = os.path.join("sqlite_data", database)

msgs = "psych290_data/dla_tutorial/msgs.csv"
csvToSQLite(msgs, database_path, "msgs")

outcomes = "psych290_data/dla_tutorial/blog_outcomes.csv"
csvToSQLite(outcomes, database_path, "outcomes")

############# 1c) Setup database connection

# loads the %%sql extension
%load_ext sql

# connects the extension to the database
from sqlalchemy import create_engine
tutorial_db_engine =  create_engine(f"sqlite:///sqlite_data/{database}.db?charset=utf8mb4")
%sql tutorial_db_engine

# set the output limit to 50
%config SqlMagic.displaylimit = 50

## PRINT FINISHED
print(" ******* LOAD FINISHED ¯\_(ツ)_/ *******")

Cloning into 'dlatk'...
remote: Enumerating objects: 6991, done.
remote: Counting objects: 100% (1076/1076), done.
remote: Compressing objects: 100% (149/149), done.
remote: Total 6991 (delta 994), reused 935 (delta 927), pack-reused 5915 (from 1)
Receiving objects: 100% (6991/6991), 62.38 MiB | 5.70 MiB/s, done.
Resolving deltas: 100% (4947/4947), done.
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.9/1.9 MB 17.2 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 88.1/88.1 kB 5.5 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 23.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.3/8.3 MB 27.6 MB/s eta 0:00:00
  Created wheel for image: filename=image-1.5.33-py2.py3-none-any.whl size=19482 sha256=dd4d5b12de9e4c82d5e2927a3bd8b8689be6292850d76a0c9dff47aaa55dfa5e
  Stored in directory: /root/.cache/pip/wheels/62/40/4

SQL Query: CREATE TABLE msgs (message_id INT, user_id INT, date VARCHAR(31), created_time VARCHAR(31), message LONGTEXT);


Importing data, reading psych290_data/dla_tutorial/msgs.csv file
Reading 10000 rows into the table...
Reading 10000 rows into the table...
Reading 10000 rows into the table...
Reading remaining 1674 rows into the table...


SQL Query: CREATE TABLE outcomes (user_id INT, gender INT, age INT, occu VARCHAR(31), sign VARCHAR(15), is_indunk INT, is_student VARCHAR(7), is_education VARCHAR(7), is_technology VARCHAR(7));


Importing data, reading psych290_data/dla_tutorial/blog_outcomes.csv file
Reading remaining 1000 rows into the table...
 ******* LOAD FINISHED ¯\_(ツ)_/ *******


### 1d) Get **`dlatk_lexica.db`** in your Google Drive!


Some dictionaries are not stored in the GitHub download of DLATK because they are not publically shared.

So, you'll have to upload a database with it to Colab when you want to work with it. We can store it in your Google Drive and access it easily.

**Follow these steps** 📋 (*only need to do this once!*)

1.  In your Google Drive make a folder called `sqlite_databases`
2.  Open [this shared folder](https://drive.google.com/drive/folders/1nxX0Qf6vd1hnNX9ywqwVsvNLoYWo62zA). Please feel free to request access.
3.  Make a copy of `dlatk_lexica.db` and put it in `sqlite_databases` in your Google Drive, so `MyDrive\sqlite_databases`.


⚠️ Make sure the file is `dlatk_lexica.db` and stored in `sqlite_databases` -- that's where the code below expects it.

### 1e) Mount Google Drive & configure the database in SQLite

Now that you have the right copy of dlatk_lexica.db stored in your Google Drive, let's connect your Drive to this Colab!

Google will ask you to allow this notebook to access your Drive--click yes and follow prompts to login and allow.

In [ ]:
# mount Google Drive &  copy database to Colab

# this connects & mounts your Google Drive to this colab space
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

# this copies dlatk_lexica.db from your Google Drive to Colab
!cp -f "/content/drive/MyDrive/sqlite_databases/dlatk_lexica.db" "sqlite_data"

Mounted at /content/drive


This code block below enables SQLite to use `dlatk_lexica.db` and `dla_tutorial.db` in the same SQL connection!

In [ ]:
# attaches the dlatk_lexica.db so tutorial_db_engine can query both databases

from IPython import get_ipython
from sqlalchemy import event

# auto‑attach the lexica db whenever tutorial_db_engine connects
@event.listens_for(tutorial_db_engine, "connect")
def _attach_lexica(dbapi_conn, connection_record):
    dbapi_conn.execute("ATTACH DATABASE 'sqlite_data/dlatk_lexica.db' AS dlatk_lexica;")

‼️ Note: In mini tutorial 5B, we go in depth on how we use Google Drive to store databases.

### (ONLY IF NEEDED: SOFT RELOAD) **If you have a "database lock" problem**

First, go to Runtime => Restart Session. Wait for that to complete. Your colab files will be preserved during this, including the DLATK install you did earlier.

Second, execute this cell:

In [ ]:
# If you face a "database locked" issue, restart the session & run this cell to get set back up!

database = "dla_tutorial"

%reload_ext sql

from sqlalchemy import create_engine
tutorial_db_engine = create_engine(f"sqlite:///sqlite_data/{database}.db?charset=utf8mb4")
dlatk_lexica_engine = create_engine(f"sqlite:///sqlite_data/dlatk_lexica.db?charset=utf8mb4")

# set the output limit to 50
%config SqlMagic.displaylimit = 50

from IPython import get_ipython
from sqlalchemy import event

# auto‑attach the lexica db whenever tutorial_db_engine connects
@event.listens_for(tutorial_db_engine, "connect")
def _attach_lexica(dbapi_conn, connection_record):
    dbapi_conn.execute("ATTACH DATABASE 'sqlite_data/dlatk_lexica.db' AS dlatk_lexica;")

%sql tutorial_db_engine

## 2) Re-extract features

Let's quickly re-extract 1grams from the `dla_tutorial` messages table!

This will take 2.5ish minutes. Go make yourself a tea?

In [ ]:
#database = "dla_tutorial" #we set that at the top! you can set it again if you like...
msgs_table = "msgs"

In [ ]:
!dlatkInterface.py \
  --corpdb {database} \
  --corptable {msgs_table} \
  --correl_field user_id \
  --add_ngrams -n 1



TopicExtractor: gensim Mallet wrapper unavailable, using Mallet directly.

-----
DLATK Interface Initiated: 2025-04-16 22:16:44
-----
Connecting to SQLite database: /content/sqlite_data/dla_tutorial
query: PRAGMA table_info(msgs)
SQL Query: DROP TABLE IF EXISTS feat$1gram$msgs$user_id
SQL Query: CREATE TABLE feat$1gram$msgs$user_id ( id INTEGER PRIMARY KEY, group_id INTEGER, feat VARCHAR(36), value INTEGER, group_norm DOUBLE)


Creating index correl_field on table:feat$1gram$msgs$user_id, column:group_id 


SQL Query: CREATE INDEX correl_field$1gram$msgs$user_id ON feat$1gram$msgs$user_id (group_id)


Creating index feature on table:feat$1gram$msgs$user_id, column:feat 


SQL Query: CREATE INDEX feature$1gram$msgs$user_id ON feat$1gram$msgs$user_id (feat)
query: PRAGMA table_info(msgs)
SQL Query: DROP TABLE IF EXISTS feat$meta_1gram$msgs$user_id
SQL Query: CREATE TABLE feat$meta_1gram$msgs$user_id ( id INTEGER PRIMARY KEY, group_id INTEGER, feat VARCHAR(16), value INTEGER, group_norm

Good!

Let's extract our first dictionaries (which needs the 1gram table in place -- that's why we re-extracted). But before we do that, let's learn about dictionaries (= lexicons).

## 3) Getting to know dictionary (= lexicon) tables

We will use the `LIWC2015` set of dictionaries, which are stored in the database `dlatk_lexica`.

Now we set up a connection between SQLite and `dlatk_lexica.db`.

In [ ]:
# connection between sqlite and dlatk_lexica
dlatk_lexica_engine = create_engine(f"sqlite:///sqlite_data/dlatk_lexica.db?charset=utf8mb4")

In [ ]:
# activates dlatk_lexica db
%sql dlatk_lexica_engine

💡 The `%sql dlatk_lexica_engine` would be roughly 🐬🐬🐬`USE database` in MySQL. So, run `%sql tutorial_db_engine` to swtich back to the `dla_tutorial` database.

⚠️ Note: if you get an error saying "other database already connected", try running `%reload_ext sql`. If that doesn't work, resort to Runtime ==> Restart Session 🙂

To see all the dictionaries in `dlatk_lexica` run this:

In [ ]:
%sqlcmd tables

+----------------------+
|         Name         |
+----------------------+
|       LIWC2015       |
|    dd_PastPreFut     |
| dd_emnlp14_ageGender |
|      dd_permaV3      |
| dd_wassa16_affectInt |
|      fb2000_cp       |
|  fb2000_freq_t50ll   |
|        labmt         |
|    mini_LIWC2015     |
|         nrc          |
|       nrc_emot       |
|       nrc_sent       |
+----------------------+

Let's see what's in the LIWC2015 table first!! That's a new table type we should know about.

In [ ]:
%sql PRAGMA table_info(LIWC2015)

Running query in 'sqlite:///sqlite_data/dlatk_lexica.db?charset=utf8mb4'

+-----+----------+-------------+---------+------------+----+
| cid |   name   |     type    | notnull | dflt_value | pk |
+-----+----------+-------------+---------+------------+----+
|  0  |    id    |     INT     |    0    |    None    | 0  |
|  1  |   term   | VARCHAR(31) |    0    |    None    | 0  |
|  2  | category | VARCHAR(15) |    0    |    None    | 0  |
|  3  |  weight  |     INT     |    0    |    None    | 0  |
+-----+----------+-------------+---------+------------+----+

In 🐬🐬🐬 MySQL it looks more confusing, as the category column is an `enum` data type. It just means that category can be one of the values from the list. It's a glorified `VARCHAR`. Let's look into the table now.

In [ ]:
%%sql

SELECT *
FROM LIWC2015
LIMIT 10;

Running query in 'sqlite:///sqlite_data/dlatk_lexica.db?charset=utf8mb4'

+----+---------+----------+--------+
| id |   term  | category | weight |
+----+---------+----------+--------+
| 1  |    he   |  PPRON   |   1    |
| 2  |   he'd  |  PPRON   |   1    |
| 3  |   he's  |  PPRON   |   1    |
| 4  |   her   |  PPRON   |   1    |
| 5  |   hers  |  PPRON   |   1    |
| 6  | herself |  PPRON   |   1    |
| 7  |   hes   |  PPRON   |   1    |
| 8  |   him   |  PPRON   |   1    |
| 9  | himself |  PPRON   |   1    |
| 10 |   his   |  PPRON   |   1    |
+----+---------+----------+--------+

In [ ]:
%%sql

SELECT term, COUNT(*) as occ
FROM LIWC2015
GROUP BY term
ORDER BY occ DESC
LIMIT 20;

Running query in 'sqlite:///sqlite_data/dlatk_lexica.db?charset=utf8mb4'

+-----------+-----+
|    term   | occ |
+-----------+-----+
|    weve   |  10 |
|   we've   |  10 |
|   we'll   |  10 |
|  weakened |  9  |
|   we're   |  9  |
|    we'd   |  9  |
|   she's   |  9  |
|   she'll  |  9  |
| perfected |  9  |
|   lowest  |  9  |
|   lowers  |  9  |
|  lowered  |  9  |
|   lower   |  9  |
|    lets   |  9  |
|   let's   |  9  |
|    ily*   |  9  |
|    hes    |  9  |
|    he's   |  9  |
|   fucks   |  9  |
|  fucked*  |  9  |
+-----------+-----+

Ah, Ok, we are getting a sense. so it lists `terms` that are in `category`s, and these terms can have `weight`s (which are all 1 for LIWC -- it's an unweighted dictionary).

#### 👩‍🔬💻 Exercise

So, can you write the command to check the number of words in each category?

⚠️ If you're using the `dla_tutorial` database, make sure to give the location of the LIWC2015 table specifically as residing in `dlatk_lexica` (e.g., `SELECT * FROM dlatk_lexica.LIWC2015`). Otherwise you have to switch into the dlatk_lexica database by running `%sql dlatk_lexica_engine` (🐬🐬🐬 `use dlatk_lexica` in MySQL).

In [ ]:
%%sql

SELECT category, count(*)
FROM LIWC2015
group by category

Running query in 'sqlite:///sqlite_data/dlatk_lexica.db?charset=utf8mb4'

+--------------+----------+
|   category   | count(*) |
+--------------+----------+
|   ACHIEVE    |   213    |
|     ADJ      |   764    |
|    ADVERB    |   140    |
|    AFFECT    |   1417   |
| AFFILIATION  |   248    |
|    ANGER     |   230    |
|     ANX      |   116    |
|   ARTICLE    |    3     |
|    ASSENT    |    36    |
|   AUXVERB    |   141    |
|     BIO      |   748    |
|     BODY     |   215    |
|    CAUSE     |   135    |
|   CERTAIN    |   113    |
|   COGPROC    |   797    |
|   COMPARE    |   317    |
|     CONJ     |    43    |
|    DEATH     |    74    |
|    DIFFER    |    81    |
|   DISCREP    |    83    |
|    DRIVES    |   1102   |
|    FAMILY    |   118    |
|     FEEL     |   128    |
|    FEMALE    |   124    |
|    FILLER    |    14    |
| FOCUSFUTURE  |    97    |
|  FOCUSPAST   |   341    |
| FOCUSPRESENT |   428    |
|    FRIEND    |    95    |
|   FUNCTION   |   491    |
|    HEALTH    |   294    |
|     HEAR     |    93    |
|     HOME     |   100    |
|      I       |    24    |
|   INFORMAL   |   388    |
|    INGEST    |   184    |
|   INSIGHT    |   259    |
|   INTERROG   |    48    |
|    IPRON     |    59    |
|   LEISURE    |   296    |
|     MALE     |   116    |
|    MONEY     |   226    |
|    MOTION    |   325    |
|    NEGATE    |    62    |
|    NEGEMO    |   746    |
|   NETSPEAK   |   217    |
|    NONFLU    |    19    |
|    NUMBER    |    36    |
|   PERCEPT    |   436    |
|    POSEMO    |   622    |
+--------------+----------+
Truncated to displaylimit of 50.

Nice!! That's looking like a good summary. Alright! Let's look at what's in the `POSEMO` dictionary. It says below there are 622 words in there.


In [ ]:
%%sql

SELECT COUNT(*)
FROM LIWC2015
WHERE category = "POSEMO";

Running query in 'sqlite:///sqlite_data/dlatk_lexica.db?charset=utf8mb4'

+----------+
| COUNT(*) |
+----------+
|   622    |
+----------+

Let's get a few random samples by executing the below query a few times.

In [ ]:
%%sql

SELECT *
FROM LIWC2015
WHERE category = "POSEMO"
ORDER BY RANDOM()
LIMIT 10;

Running query in 'sqlite:///sqlite_data/dlatk_lexica.db?charset=utf8mb4'

+------+-------------+----------+--------+
|  id  |     term    | category | weight |
+------+-------------+----------+--------+
| 909  |   fondness  |  POSEMO  |   1    |
| 1301 |   usefully  |  POSEMO  |   1    |
| 969  |  handsomest |  POSEMO  |   1    |
| 797  | comfortable |  POSEMO  |   1    |
| 1147 |  pleasant*  |  POSEMO  |   1    |
| 1216 |    smart    |  POSEMO  |   1    |
| 1185 |  respected  |  POSEMO  |   1    |
| 1272 |   thankful  |  POSEMO  |   1    |
| 790  |    cheers   |  POSEMO  |   1    |
| 743  |  beautiful  |  POSEMO  |   1    |
+------+-------------+----------+--------+

Alrighty! 622 words, some of those stemmed (with `*` asterixes). DLATK will know what to do with it when we extract the dictionary -- it will match all the words that start with this stem.

#### 👩‍🔬💻 Exercise

Also, let's check some terms are mentioned in more than one dictionary category.

⚠️ Make sure to use right database!

In [ ]:
%%sql

SELECT term, count(*) as number_of_categories
FROM LIWC2015
GROUP BY term
ORDER BY number_of_categories DESC;

Running query in 'sqlite:///sqlite_data/dlatk_lexica.db?charset=utf8mb4'

+-----------+----------------------+
|    term   | number_of_categories |
+-----------+----------------------+
|    weve   |          10          |
|   we've   |          10          |
|   we'll   |          10          |
|  weakened |          9           |
|   we're   |          9           |
|    we'd   |          9           |
|   she's   |          9           |
|   she'll  |          9           |
| perfected |          9           |
|   lowest  |          9           |
|   lowers  |          9           |
|  lowered  |          9           |
|   lower   |          9           |
|    lets   |          9           |
|   let's   |          9           |
|    ily*   |          9           |
|    hes    |          9           |
|    he's   |          9           |
|   fucks   |          9           |
|  fucked*  |          9           |
|    fuck   |          9           |
|   youve   |          8           |
|   youre   |          8           |
|   youll   |          8           |
|   you've  |          8           |
|   you're  |          8           |
|   you'll  |          8           |
|   wholl   |          8           |
|   who'll  |          8           |
|  weakest  |          8           |
|   weaker  |          8           |
|  weakens  |          8           |
|  twitchy  |          8           |
|  trusted  |          8           |
|   theyve  |          8           |
|   theyll  |          8           |
|  they've  |          8           |
|  they'll  |          8           |
|    shes   |          8           |
|   she'd   |          8           |
|  perfect  |          8           |
|  lowering |          8           |
|   lover*  |          8           |
|   loses   |          8           |
|    lose   |          8           |
|   leads   |          8           |
|  leading  |          8           |
|    lead   |          8           |
|  inferior |          8           |
| inadequa* |          8           |
+-----------+----------------------+
Truncated to displaylimit of 50.

Alright, seem's like some pronouns (+ stems) are in multiple dictionaries. But let's see what categories the word `perfected` (which is a content word) is in.

#### 👩‍🔬💻 Exercise

Can you write the command to check the categories that `perfected` is in?

In [ ]:
%%sql

SELECT term, category
FROM LIWC2015
WHERE term = "perfected"

Running query in 'sqlite:///sqlite_data/dlatk_lexica.db?charset=utf8mb4'

+-----------+-----------+
|    term   |  category |
+-----------+-----------+
| perfected |   POSEMO  |
| perfected |   AFFECT  |
| perfected |  COGPROC  |
| perfected | FOCUSPAST |
| perfected |   DRIVES  |
| perfected |    VERB   |
| perfected |  ACHIEVE  |
| perfected |   REWARD  |
| perfected |  CERTAIN  |
+-----------+-----------+

Interesting...

This tells us something about LIWC -- the same word can show up all over the place.

FYI, some of these categories are nested: every word in `POSEMO` is also in the `AFFECT` dictionary, all `ACHIEVE` words are in `DRIVE`, etc. (see page 4 here in the [LIWC manual](https://repositories.lib.utexas.edu/bitstream/handle/2152/31333/LIWC2015_LanguageManual.pdf)).

While we are at it, let's look at the word "love", which does all sorts of damage when measured with dictionaries [(see here for a discussion](https://static1.squarespace.com/static/53d29678e4b04e06965e9423/t/5f2ad5ec2985cb6e8e2e6548/1596642799449/PNAS-2020-subjectiveWellBeing.pdf), third page, under Highly Frequent Words).

In [ ]:
%%sql

SELECT term, category
FROM LIWC2015
WHERE term = "love"

Running query in 'sqlite:///sqlite_data/dlatk_lexica.db?charset=utf8mb4'

+------+-------------+
| term |   category  |
+------+-------------+
| love |    POSEMO   |
| love |    AFFECT   |
| love |     BIO     |
| love |    SOCIAL   |
| love |    DRIVES   |
| love | AFFILIATION |
+------+-------------+

Good to know where it shows up!

## 4) Dictionary extraction

Alright, we've gotten a sense of how dictionary tables work. Let's extract dictionaries. Basically, here is the logic

* For every user (i.e., group_id)
    * For every dictionary
        * Count how many words that he/she uses are in the dictionary
        * Divide that number by the total number of words that he/she has written

This will be in a feature table which will tell us, for example, that `5.3%` of the words by user `11111` are `POSEMO` words (that is, match `terms` in the POSEMO `category` (dictionary) in the LIWC2015 table (set of dictionaries)).

Ok, let's tell DLATK to do this for all users with a mini LIWC2015 (`mini_LIWC2015`) set of dictionaries that's a little easier to explore. It just contains `POSEMO`, `NEGEMO` and `SOCIAL` words.

In [ ]:
%%sql

SELECT category, COUNT(*) AS terms_in_category
FROM mini_LIWC2015
GROUP BY category;

Running query in 'sqlite:///sqlite_data/dlatk_lexica.db?charset=utf8mb4'

+----------+-------------------+
| category | terms_in_category |
+----------+-------------------+
|  NEGEMO  |        746        |
|  POSEMO  |        622        |
|  SOCIAL  |        756        |
+----------+-------------------+

### 4a) DLATK command: Dictionary extraction

⚠️ Before extracting dictionary features, we need to make sure that we have the 1gram feature table as dictionary extraction depends on it. Else, dictionary extraction throws an error.

Other than the first three, that we always need for a DLATK command, the dictionary extraction flags are

`--add_lex_table -l mini_LIWC2015`,

where the string after -l gives the name of the dictionary table in the `dlatk_lexica` database we want to extract. In this case, it's `mini_LIWC2015`

⚠️ note: long flags get two dashes (`--add_lex_table`), short single letter flags and parameters are given with one dash (`-l mini_LIWC2015`, or `-n 1`).

In [ ]:
database = "dla_tutorial"
msgs_table = "msgs"

In [ ]:
!dlatkInterface.py \
    --corpdb {database} \
    --corptable {msgs_table} \
    --correl_field user_id \
    --add_lex_table -l mini_LIWC2015

That was quick. So first of all, let's look at the new feature table that was created here:

`feat$cat_mini_LIWC2015$msgs$user_id$1gra`

## 5) Understanding dictionary feature tables

Let's look at the columns in the feature table `feat$cat_mini_LIWC2015$msgs$user_id$1gra`.

First we point our handy `%sql` extension to the dla_tutorial database again.

⚠️ Remember, with this database switching, you may run into an error like "other database already being used!", then you just `run %reload_ext sql`!

In [ ]:
# %reload_ext sql # if needed!

In [ ]:
# switch into tutorial_db_engine (= "dla_tutorial.db")
%sql tutorial_db_engine

In [ ]:
feat_mini_liwc_usr = 'feat$cat_mini_LIWC2015$msgs$user_id$1gra'

In [ ]:
%%sql

PRAGMA table_info( {{feat_mini_liwc_usr}} )

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+-----+------------+-------------+---------+------------+----+
| cid |    name    |     type    | notnull | dflt_value | pk |
+-----+------------+-------------+---------+------------+----+
|  0  |     id     |   INTEGER   |    0    |    None    | 1  |
|  1  |  group_id  |   INTEGER   |    0    |    None    | 0  |
|  2  |    feat    | VARCHAR(10) |    0    |    None    | 0  |
|  3  |   value    |   INTEGER   |    0    |    None    | 0  |
|  4  | group_norm |    DOUBLE   |    0    |    None    | 0  |
+-----+------------+-------------+---------+------------+----+

```
🐬🐬🐬
SHOW COLUMNS FROM {feat_mini_liwc_usr};
🐬🐬🐬
```

Feature tables can be confusing. The thing to note about them is  
* their columns/fields always have the same names (`id`, `group_id`, `feat`, `value`, `group_norm`).
* the **name** of the feature table tells us how the table came about, and what exactly is in it.

In the name of this table, the `cat_` prefix means that it was created with `--add_lex_table`.

Let's look at the contents now. The offset command in conjunction with the limit skips the first 100 rows, and then shows you 10. It's a way to avoid always looking at the same darn 10 rows at the top of the table (instead you are now looking at rows 101-110).

In [ ]:
feat_mini_liwc_usr = 'feat$cat_mini_LIWC2015$msgs$user_id$1gra'

In [ ]:
%%sql

SELECT *
FROM {{feat_mini_liwc_usr}}
LIMIT 10
OFFSET 100;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+-----+----------+------------+-------+-----------------------+
|  id | group_id |    feat    | value |       group_norm      |
+-----+----------+------------+-------+-----------------------+
| 101 |  911744  |   POSEMO   |   28  |  0.019138755980861254 |
| 102 |  911744  |   SOCIAL   |   86  |  0.05878332194121671  |
| 103 |  911744  |   NEGEMO   |   8   | 0.0054682159945317835 |
| 104 |  911744  | _intercept |   1   |          1.0          |
| 105 |  918175  |   NEGEMO   |   83  |  0.01651084145613686  |
| 106 |  918175  |   SOCIAL   |  433  |  0.08613487169285848  |
| 107 |  918175  |   POSEMO   |   87  |  0.017306544658842248 |
| 108 |  918175  | _intercept |   1   |          1.0          |
| 109 |  942828  |   POSEMO   | 11500 |  0.03148002660746589  |
| 110 |  942828  |   SOCIAL   | 33033 |  0.09042432338473333  |
+-----+----------+------------+-------+-----------------------+

In table `feat$cat_mini_LIWC2015$msgs$user_id$1gra`,
- **group_id** -- `user_id` from `msgs`, represents a user
- **feat** -- One of several categories (dictionaries) from our lexicon table, `mini_LIWC2015`
- **value** -- Number of times a word in the category mentioned in `feat` is used by user in `group_id`
- **group_norm** -- Proportion of words used by user in `group_id` that belong to category in `feat`

### 5a) How the feature table name connects to its contents

This is now the second time that we are encountering a feature table -- in the previous tutorial, we had a 1gram table. Now we have a dictionary table.

For the dictionary table `feat$cat_mini_LIWC2015$msgs$user_id$1gra`

| this column in the feature table... | ...contains this (as recorded in table name)|
|------|------|
| feat | **categories in _mini_LIWC2015_** |
| group_id | user_id |

Compare this to the 1gram table `feat$1gram$msgs$user_id$16to16`

| this column in the feature table... | ...contains this (as recorded in table name)|
|------|------|
| feat | **1gram** |
| group_id | user_id |

For example, for a particular user_id:

In [ ]:
feat_mini_liwc_usr = 'feat$cat_mini_LIWC2015$msgs$user_id$1gra'

In [ ]:
%%sql

SELECT *
FROM {{feat_mini_liwc_usr}}
WHERE group_id = 911744;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+-----+----------+------------+-------+-----------------------+
|  id | group_id |    feat    | value |       group_norm      |
+-----+----------+------------+-------+-----------------------+
| 101 |  911744  |   POSEMO   |   28  |  0.019138755980861254 |
| 102 |  911744  |   SOCIAL   |   86  |  0.05878332194121671  |
| 103 |  911744  |   NEGEMO   |   8   | 0.0054682159945317835 |
| 104 |  911744  | _intercept |   1   |          1.0          |
+-----+----------+------------+-------+-----------------------+

User `911744` has used:
- 28 words in LIWC category `POSEMO`
- 86 words in LIWC category `SOCIAL`
- 8 words in LIWC category `NEGEMO`
- `_intercept`: **you can ignore this, if you like**. It's a dummy for every `group_id` that DLATK adds to make sure that every `group_id` shows up in this table, even if a particular `group_id` did not use any words in the `feat` categories (here: POSEMO, SOCIAL, or NEGEMO). This is to take into account the sparse encoding of feature tables-- group_ids would not get feature rows for dictionaries they did not mention, and so would not appear in the table otherwise.

Having this `_intercept` in here makes sure that if you run `count(distinct(group_id))` on a lexicon table,  you'll always get the total number of `group_id`'s that were in the message table.

At this point, a LIWC-minded person might say this person expresses more social words than negemo words, for example (but that makes strong assumptions about the recall of the dictionaries -- that they really hit all the social and negemo expressions that are out there in the world).

#### 👩‍🔬💻 Exercise

Do the group_norms (ignoring `_intercept`) sum to 1 in a lex table? Why or why not?

In [ ]:
feat_mini_liwc_usr = 'feat$cat_mini_LIWC2015$msgs$user_id$1gra '

In [ ]:
%%sql

SELECT group_id, SUM(group_norm)
FROM (SELECT *
      FROM {{feat_mini_liwc_usr}}
      WHERE feat != '_intercept') AS without_inter
GROUP BY group_id;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+----------+----------------------+
| group_id |   SUM(group_norm)    |
+----------+----------------------+
|  28451   | 0.12488083889418486  |
|  174357  | 0.15716486902927582  |
|  216833  |  0.1058989617398455  |
|  317581  | 0.13893232715951898  |
|  446275  |  0.1562216167120802  |
|  450169  | 0.07966457023060797  |
|  468786  | 0.18045977011494235  |
|  477665  | 0.14034593850677132  |
|  483266  | 0.08835904628330996  |
|  485307  |  0.1370772946859904  |
|  517266  | 0.13261142498430661  |
|  518116  | 0.15179005783679617  |
|  664485  | 0.14857738158838873  |
|  669116  | 0.12931151907738356  |
|  671748  | 0.10703414918900878  |
|  673235  | 0.14161849710982669  |
|  715237  | 0.12890972278348073  |
|  734023  | 0.12008281573498963  |
|  743739  | 0.10669456066945596  |
|  791840  | 0.11521367521367525  |
|  808956  |  0.1239151179170268  |
|  825029  | 0.12427716849451619  |
|  828778  |  0.130071099958176   |
|  832265  | 0.10581094654955064  |
|  882528  |  0.1475457449898346  |
|  911744  | 0.08339029391660975  |
|  918175  |  0.1199522578078376  |
|  942828  | 0.13723101685960792  |
|  950122  | 0.11303524958599523  |
|  979795  | 0.11027036594539658  |
| 1007188  | 0.08979841172877218  |
| 1011311  |  0.1315840419353401  |
| 1039908  | 0.12379807692307683  |
| 1077954  | 0.07826396300249026  |
| 1111991  | 0.15734767025089585  |
| 1143104  | 0.16497203863751939  |
| 1156278  | 0.09187089862059691  |
| 1188176  | 0.12980623215695727  |
| 1196970  |  0.1230095841823261  |
| 1197086  | 0.061083743842364514 |
| 1223044  | 0.13636363636363635  |
| 1223561  | 0.11499360808425114  |
| 1225881  | 0.09649871904355249  |
| 1232458  |  0.1354252683732452  |
| 1234212  | 0.12089731544657903  |
| 1240274  | 0.12202556013047483  |
| 1254400  | 0.14675052410901468  |
| 1261679  | 0.13554216867469912  |
| 1270372  |  0.0978027867095392  |
| 1292625  |  0.1346079753419379  |
+----------+----------------------+
Truncated to displaylimit of 50.

As shown above, the `group_norm` without `_intercept` don't sum up to one, because all 1-grams need not occur in at least one dictionary.

In [ ]:
feat_1gram_usr = 'feat$1gram$msgs$user_id'

In [ ]:
%%sql

SELECT group_id, SUM(value) AS tokens
FROM {{feat_1gram_usr}}
GROUP BY group_id
ORDER BY group_id ASC;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+----------+--------+
| group_id | tokens |
+----------+--------+
|  28451   |  1049  |
|  174357  |  649   |
|  216833  | 55959  |
|  317581  | 69703  |
|  446275  |  2202  |
|  450169  |  954   |
|  468786  |  870   |
|  477665  | 31971  |
|  483266  |  713   |
|  485307  |  1656  |
|  517266  |  6372  |
|  518116  | 111002 |
|  664485  | 262052 |
|  669116  | 66073  |
|  671748  | 47966  |
|  673235  |  2076  |
|  715237  | 12373  |
|  734023  |  966   |
|  743739  |  3824  |
|  791840  |  2925  |
|  808956  | 22007  |
|  825029  | 40120  |
|  828778  | 21519  |
|  832265  | 46791  |
|  882528  |  6886  |
|  911744  |  1463  |
|  918175  |  5027  |
|  942828  | 365311 |
|  950122  | 21135  |
|  979795  | 135594 |
| 1007188  |  1637  |
| 1011311  | 67151  |
| 1039908  |  7488  |
| 1077954  |  2811  |
| 1111991  |  5580  |
| 1143104  |  7868  |
| 1156278  | 29868  |
| 1188176  | 32926  |
| 1196970  | 26502  |
| 1197086  |  1015  |
| 1223044  |  2464  |
| 1223561  | 16427  |
| 1225881  |  1171  |
| 1232458  |  1211  |
| 1234212  | 259708 |
| 1240274  | 18701  |
| 1254400  |  477   |
| 1261679  | 24236  |
| 1270372  |  7464  |
| 1292625  | 20764  |
+----------+--------+
Truncated to displaylimit of 50.

In [ ]:
feat_mini_liwc_usr = 'feat$cat_mini_LIWC2015$msgs$user_id$1gra '

In [ ]:
%%sql

SELECT group_id, SUM(value) AS tokens
FROM {{feat_mini_liwc_usr}}
GROUP BY group_id
ORDER BY group_id ASC;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+----------+--------+
| group_id | tokens |
+----------+--------+
|  28451   |  132   |
|  174357  |  103   |
|  216833  |  5927  |
|  317581  |  9685  |
|  446275  |  345   |
|  450169  |   77   |
|  468786  |  158   |
|  477665  |  4488  |
|  483266  |   64   |
|  485307  |  228   |
|  517266  |  846   |
|  518116  | 16850  |
|  664485  | 38936  |
|  669116  |  8545  |
|  671748  |  5135  |
|  673235  |  295   |
|  715237  |  1596  |
|  734023  |  117   |
|  743739  |  409   |
|  791840  |  338   |
|  808956  |  2728  |
|  825029  |  4987  |
|  828778  |  2800  |
|  832265  |  4952  |
|  882528  |  1017  |
|  911744  |  123   |
|  918175  |  604   |
|  942828  | 50133  |
|  950122  |  2390  |
|  979795  | 14953  |
| 1007188  |  148   |
| 1011311  |  8837  |
| 1039908  |  928   |
| 1077954  |  221   |
| 1111991  |  879   |
| 1143104  |  1299  |
| 1156278  |  2745  |
| 1188176  |  4275  |
| 1196970  |  3261  |
| 1197086  |   63   |
| 1223044  |  337   |
| 1223561  |  1890  |
| 1225881  |  114   |
| 1232458  |  165   |
| 1234212  | 31399  |
| 1240274  |  2283  |
| 1254400  |   71   |
| 1261679  |  3286  |
| 1270372  |  731   |
| 1292625  |  2796  |
+----------+--------+
Truncated to displaylimit of 50.


Let's look at users who have the largest proportions of `POSEMO`. For proportions we have to look at the
`group_norm` field.

#### 👩‍🔬💻 Exercise

Can you now guess the SQL spell to show which users have the largest proportion of LIWC `POSEMO` words?

In [ ]:
feat_mini_liwc_usr = 'feat$cat_mini_LIWC2015$msgs$user_id$1gra'

In [ ]:
%%sql

SELECT group_id AS user_id, feat, group_norm
FROM {{feat_mini_liwc_usr}}
WHERE FEAT = "POSEMO"
ORDER BY group_norm DESC
limit 10;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+---------+--------+---------------------+
| user_id |  feat  |      group_norm     |
+---------+--------+---------------------+
| 3911122 | POSEMO |  0.0921273031825796 |
| 3677344 | POSEMO | 0.08806262230919765 |
| 3656035 | POSEMO | 0.07001795332136444 |
| 3319097 | POSEMO |  0.0687582190494082 |
| 3901750 | POSEMO | 0.06677631578947363 |
| 3999305 | POSEMO |  0.0664546899841016 |
| 3723475 | POSEMO | 0.06494960806270997 |
| 3811749 | POSEMO | 0.06389776357827476 |
|  174357 | POSEMO | 0.06317411402157168 |
| 4122908 | POSEMO | 0.06194690265486724 |
+---------+--------+---------------------+

Now, let's get the average `group_norm` for each LIWC category.

In [ ]:
%%sql

SELECT feat, AVG(group_norm)
FROM {{feat_mini_liwc_usr}}
GROUP BY feat
limit 10;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+------------+----------------------+
|    feat    |   AVG(group_norm)    |
+------------+----------------------+
|   NEGEMO   | 0.018002871887595064 |
|   POSEMO   | 0.030697855849891553 |
|   SOCIAL   | 0.07696711485927378  |
| _intercept |         1.0          |
+------------+----------------------+

`POSEMO` is a bit less than twice as frequent than `NEGEMO` overall. And `SOCIAL` is the largest.

BTW, here is a way to get more statistics, all at once:

In [ ]:
%%sql

SELECT feat, AVG(group_norm), MIN(group_norm), MAX(group_norm), count(distinct(group_id)) AS included_users
FROM {{feat_mini_liwc_usr}}
GROUP BY feat;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+------------+----------------------+-----------------------+---------------------+----------------+
|    feat    |   AVG(group_norm)    |    MIN(group_norm)    |   MAX(group_norm)   | included_users |
+------------+----------------------+-----------------------+---------------------+----------------+
|   NEGEMO   | 0.018002871887595064 | 0.0007887990534411359 | 0.05504587155963303 |      998       |
|   POSEMO   | 0.030697855849891553 |  0.002593473092716663 |  0.0921273031825796 |      1000      |
|   SOCIAL   | 0.07696711485927378  |  0.004281207751239298 |  0.1864951768488747 |      1000      |
| _intercept |         1.0          |          1.0          |         1.0         |      1000      |
+------------+----------------------+-----------------------+---------------------+----------------+

it's good to include the "included_users" column to remind ourselves that we are doing "dumb" group_norm averages -- generally an acceptable strategy when it comes to dictionaries with decent coverage. Apparently there were two users who have never used a NEGEMO word in the dataset -- good for theeeeeem.

## 6) Which words are in a dictionary(s), and drive its frequency?

As discussed in the lectures, we need to know what's going on inside a dictionary, we can't really take it at face value. For that we have to know which words drive its occurrence.

For that, we have to combine two pieces of information:

* Table `feat$1gram$msgs$user_id` contains the number of occurrences of tokens (words) for users
* Table `mini_LIWC2015` contains which words belong to a LIWC category.

Let us merge the two on words (`feat` in the 1gram table, `term` in the lexicon table) so we can find out for each category, which words occurred the most.

This will show us which words are dominant in a category for our corpus.

### 6a) Step 1: make a word count table -- counts per words

The word counts in `feat$1gram$msgs$user_id` are per user (that's the group_id), and the meta_tables (if you thought of those) too report statistics *per user*. We want total counts per words.


#### 👩‍🔬💻 Exercise

Can you make a new table `word_counts` that will have those?


In [ ]:
feat_1gram_usr = 'feat$1gram$msgs$user_id'

In [ ]:
%%sql

DROP TABLE IF EXISTS word_counts;

CREATE TABLE word_counts AS SELECT feat AS word, SUM(value) AS count
                             FROM {{feat_1gram_usr}}
                             GROUP BY feat;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

++
||
++
++

Great, this gives us a table with overall word counts. Let's just check what are the most frequent ones.

In [ ]:
%%sql

SELECT *
FROM word_counts
ORDER BY count DESC
LIMIT 10;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+------+--------+
| word | count  |
+------+--------+
|  .   | 365156 |
|  ,   | 297114 |
| the  | 265240 |
|  i   | 224789 |
|  to  | 193945 |
| and  | 182691 |
|  a   | 138493 |
|  of  | 121843 |
| that | 86253  |
|  in  | 85981  |
+------+--------+

### 6b) Step 2: merge on the dictionary table

Alright! These words make sense. Now, let's filter this table down to only the words in the `mini_LIWC2015` dict.

The `a.*` below is clever -- it tells SQL to show us all columns contained in the first table (and then whatever we want from table b -- `b.*` would also be an option.) Make sure you understand the above query. `where a.word = b.term` with two tables is the same as an inner join (it's an **Implicit Inner Join** to be precise).

In [ ]:
%%sql

SELECT a.*, b.category
FROM word_counts AS a, dlatk_lexica.mini_LIWC2015 AS b
WHERE a.word = b.term
ORDER BY count DESC
LIMIT 10;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+------+-------+----------+
| word | count | category |
+------+-------+----------+
| you  | 58284 |  SOCIAL  |
|  we  | 34652 |  SOCIAL  |
|  he  | 29493 |  SOCIAL  |
| they | 24339 |  SOCIAL  |
| she  | 18642 |  SOCIAL  |
| her  | 17899 |  SOCIAL  |
| his  | 16757 |  SOCIAL  |
| well | 14929 |  POSEMO  |
| your | 14505 |  SOCIAL  |
| who  | 14413 |  SOCIAL  |
+------+-------+----------+

🤔 that took a weebit long... I wonder why that is? do you know what we could have done to make it faster?

In [ ]:
%%sql

CREATE INDEX idx_word_counts_word
ON word_counts(word);

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

++
||
++
++

```
🐬🐬🐬
ALTER TABLE word_counts ADD index (word);
🐬🐬🐬
```

Now run the cell again that does the implicit join above the one that created the index just now! Faster?? Seeeee-- indices! Like magic, except less exciting.

Let's save the join as a new table by wrapping a `create table ` around it.

In [ ]:
%%sql

DROP TABLE IF EXISTS miniLiwc_wordcounts;

CREATE TABLE miniLiwc_wordcounts AS SELECT a.*, b.category
                                    FROM word_counts AS a, dlatk_lexica.mini_LIWC2015 AS b
                                    WHERE a.word = b.term
                                    ORDER BY count DESC;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

++
||
++
++

### 6c) Investigating different categories (sub-dictionaries) within the (dlatk_lexica) dictionary

Ok, cool! So now let's only look at the `POSEMO` category -- what words occur the most in it?

In [ ]:
%%sql

SELECT *
FROM miniLiwc_wordcounts
WHERE category = "POSEMO"
ORDER BY count DESC
limit 10;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+--------+-------+----------+
|  word  | count | category |
+--------+-------+----------+
|  well  | 14929 |  POSEMO  |
|  good  | 13161 |  POSEMO  |
|  love  |  9346 |  POSEMO  |
| great  |  5306 |  POSEMO  |
| better |  4721 |  POSEMO  |
|  fun   |  4165 |  POSEMO  |
| pretty |  4056 |  POSEMO  |
|  best  |  3924 |  POSEMO  |
|  nice  |  3861 |  POSEMO  |
|  hope  |  3688 |  POSEMO  |
+--------+-------+----------+

Turns out, `well`, `good`, `love` are all bad words in that they are highly ambiguous. [see here](https://static1.squarespace.com/static/53d29678e4b04e06965e9423/t/5f2ad5ec2985cb6e8e2e6548/1596642799449/PNAS-2020-subjectiveWellBeing.pdf)

Can you mod the query above to show you the POSEMO words only used once?

Let's check in on the other two as well.

In [ ]:
%%sql

SELECT *
FROM miniLiwc_wordcounts
WHERE category = 'NEGEMO'
ORDER BY count DESC
LIMIT 10;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+--------+-------+----------+
|  word  | count | category |
+--------+-------+----------+
|  bad   |  4533 |  NEGEMO  |
|  hate  |  2394 |  NEGEMO  |
| wrong  |  2286 |  NEGEMO  |
|  lost  |  2146 |  NEGEMO  |
|  hell  |  2071 |  NEGEMO  |
| stupid |  1889 |  NEGEMO  |
|  miss  |  1753 |  NEGEMO  |
| sorry  |  1696 |  NEGEMO  |
|  fuck  |  1690 |  NEGEMO  |
| alone  |  1664 |  NEGEMO  |
+--------+-------+----------+

In [ ]:
%%sql

SELECT *
FROM miniLiwc_wordcounts
WHERE category = 'SOCIAL'
ORDER BY count DESC
LIMIT 10;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+------+-------+----------+
| word | count | category |
+------+-------+----------+
| you  | 58284 |  SOCIAL  |
|  we  | 34652 |  SOCIAL  |
|  he  | 29493 |  SOCIAL  |
| they | 24339 |  SOCIAL  |
| she  | 18642 |  SOCIAL  |
| her  | 17899 |  SOCIAL  |
| his  | 16757 |  SOCIAL  |
| your | 14505 |  SOCIAL  |
| who  | 14413 |  SOCIAL  |
| him  | 13003 |  SOCIAL  |
+------+-------+----------+

Who would have guessed that `SOCIAL` is full of function word pronouns!? It's basically just a pronoun dictionary, the rest won't matter. Good we checked!!

Bahh! Ignorance is bliss, except, of course, it isn't.

## 7) Message-level extraction: Blog posts that have POSEMO words

During the lectures, we saw that to judge the precision of a dictionary, we need to look at (1) of those messages containing dictionary words (2) how many actually express the concept conveyed by the dictionary. Specifically (2)/(1) is "precision."

Lets find some blogs with POSEMO words in it -- so that we can look at them, with our eyes, and see if they really are positive emotion-y.

For that, we again want to match the words in the POSEMO category against a feature table that contains words -- so 1grams. But now, we don't want to group blog posts from users, we want to get the words in the specific blog posts, so we can fish out blog posts that have POSEMO words in them.

DLATK will come to the rescue -- through the clever use of the `correl_field` (=group-by_field) -- we will tell it to extract 1-grams with `message_id` as our `correl_field`. That will get us a `feat$1gram$msgs$message_id` feature table.

Then we find `message_id` of `feat` in `feat$1gram$msgs$message_id` that belong `term` in `category` POSEMO in `mini_LIWC2015`.

This will take a ~25 minutes!

In [ ]:
!dlatkInterface.py \
  --corpdb {database} \
  --corptable {msgs_table} \
  --correl_field message_id \
  --add_ngrams -n 1

Just to remind ourselves again --
* all feature tables have the same columns (`id`, `group_id`, `feat`, `value`, `group_norm`).
* the **name** of the feature table tells us how the table came about, and what exactly is in it.

### 7a) A feature table for messages!

This is now the third time that we are encountering a feature table.

Our new table is: `feat$1gram$msgs$message_id`

| this column in the feature table... | ...contains this (as recorded in table name)|
|------|------|
| feat | 1gram |
| group_id | message_id |

**Note that the group_id field now is no longer users, but messages (here: blog posts. That's what's in the message table).**

Previously we had, just for comparison:

The dictionary table `feat$cat_mini_LIWC2015$msgs$user_id$1gra`

| this column in the feature table... | ...contains this (as recorded in table name)|
|------|------|
| feat | **categories in _mini_LIWC2015_** |
| group_id | user_id |

And the 1gram table `feat$1gram$msgs$user_id`

| this column in the feature table... | ...contains this (as recorded in table name)|
|------|------|
| feat | **1gram** |
| group_id | user_id |

Hopefully you are getting the idea. It's an abstract data structure 🌈

Let's sanity check the new table. How many messages does it contain features for?




In [ ]:
feat_1gram_msg = 'feat$1gram$msgs$message_id'

In [ ]:
%%sql

SELECT COUNT(DISTINCT(group_id)) AS num_msgs
FROM {{feat_1gram_msg}};

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+----------+
| num_msgs |
+----------+
|  31502   |
+----------+

So yeah, it's a big table! Because it has 31,000 messages times all their 1gram counts, for a total of how many rows?

In [ ]:
feat_1gram_msg = 'feat$1gram$msgs$message_id'

In [ ]:
%%sql

SELECT COUNT(*) AS num_rows
FROM {{feat_1gram_msg}};

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+----------+
| num_rows |
+----------+
| 3764873  |
+----------+

3.7m rows -- we are not in Excel Land anymore.

Let's keep swimming though.

Now we want to create a query that pulls out the `message_id`s that have 1grams that match the `term` column for POSEMO words.

In [ ]:
feat_1gram_msg = 'feat$1gram$msgs$message_id'

In [ ]:
%%sql

SELECT group_id AS message_id, feat
FROM feat$1gram$msgs$message_id
WHERE feat IN (SELECT term
               FROM dlatk_lexica.mini_LIWC2015
               WHERE category = 'POSEMO')
LIMIT 10;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+------------+------+
| message_id | feat |
+------------+------+
|    2523    |  (:  |
|    2526    |  (:  |
|    2750    |  (:  |
|    5896    |  (:  |
|    5912    |  (:  |
|    5922    |  (:  |
|    6000    |  (:  |
|    7298    |  (:  |
|    7381    |  (:  |
|    7382    |  (:  |
+------------+------+

So this tells us that message 2523 had the feature "(:" in it, which is in the POSEMO dictionary. Hopefully you get the idea. So let's make ourselves a table that preserves this.

In [ ]:
feat_1gram_msg = 'feat$1gram$msgs$message_id'

In [ ]:
%%sql

DROP TABLE IF EXISTS posemo_blogs;

CREATE TABLE posemo_blogs AS SELECT group_id AS message_id, feat
                              FROM {{feat_1gram_msg}}
                              WHERE feat IN (SELECT term
                                             FROM dlatk_lexica.mini_LIWC2015
                                             WHERE category = 'POSEMO');

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

++
||
++
++

and let's pull some of those messages for a random sample.

In [ ]:
%%sql

SELECT a.feat AS posemo_word, b.*
FROM posemo_blogs AS a, msgs AS b
WHERE a.message_id = b.message_id
ORDER BY RANDOM()
LIMIT 3;

Running query in 'sqlite:///sqlite_data/dla_tutorial.db?charset=utf8mb4'

+-------------+------------+---------+----------------+---------------------+-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Now we could annotate this to see if the `POSEMO` token really designates positive emotion for a random sample.

## 8) ‼️ Save your database ‼️

The homework assumes that you have extracted feature tables. So let's write a copy of your database to Google Drive! 📝

Google will ask you to allow this notebook to access your Drive--click yes and follow prompts to login and allow!

In [ ]:
database = "dla_tutorial"

In [ ]:
# Save your database in Google Drive

# mount Google Drive (if you haven't already)
from google.colab import drive
drive.mount('/content/drive', force_remount=True)

# copy the database file to your Drive (`-f` forces it to write over the old database with any changes)
!cp -f "sqlite_data/{database}.db" "/content/drive/MyDrive/sqlite_databases/"

print(f"Database has been copied to your Google Drive with success!")

Now your database is saved in your Google Drive! We can double check it's there by running this:

In [ ]:
!ls -lh "/content/drive/MyDrive/sqlite_databases"

🧐 Next up, check out the mini tutorial for more practice on saving databases to Google Drive!